### PR TITLE
FEATURE: Add commands `media:listVariantPresets` and `media:removeVariants`

### LINKED ISSUE #3086: Missing clearvariants command
If variantsPresets are not defined in the Settings anymore, there is no way to remove them. A command equivalent to clearthumbnails should be added to remove unused variants

### PR DESCRIPTION
**Pull request by Jan3k3y**
Currently there is an option to create custom ImageVariant out of the `Settings.Neos.Media.yaml`. These VariantPresets can be edited, but variants with old/removed/outdated presets will not be deleted.

To solve this problem, I wrote a few lines of code and added two new commands...

1. list all found configurations for variants - from file and database, to get every single variant configuration, still existing or not
2. delete a variant preset by identifier and name provided by the other command

ISSUE: https://github.com/neos/neos-development-collection/issues/3086

**Upgrade instructions**

Nothing I know about so far

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.Media/Classes/Domain/Service/ImageVariantService.php
Comment (kitsunet) [inline Neos.Media/Classes/Domain/Service/ImageVariantService.php]: Nitpicky, but we are commited to use as few shorts as possible, so I would prefer Configurations instead of Configs even if obivously in the context it's very clear what is meant here. Makes searching easier though.

And instead of the optional argument I would prefer a second method with two required parameters, so you can either (naming change suggetion) `getAllPresetsFromConfiguration` `getPresetFromConfigurationByName`

File: Neos.Media/Tests/Unit/Domain/Service/ImageVariantServiceTest.php
Comment (kitsunet) [inline Neos.Media/Tests/Unit/Domain/Service/ImageVariantServiceTest.php]: Configs again

File: Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php
Comment (kitsunet) [inline Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php]: You actually expect a dictionary/map here so this might be misleading, as the key is quite important and it cannot be a simple array. I think it would be good to use the `\Neos\Media\Domain\ValueObject\Configuration\Variant` and VariantPreset value objects to move those Presets around. Maybe add a PresetCollection even.

File: Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php
Comment (kitsunet) [inline Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php]: This is unclear to me, maybe I need a coffee to understand the use of this, it seems to contradict the method name to not filter?

**Review by kitsunet — COMMENTED** [review]
This is a cool addition, I left some nitpicks to go over, but it's on a good track. Thanks a lot!

**Comment by Jan3k3y** [comment]
I have adjusted some things and hope that it is now more understandable and fits to your naming commitment.

**Comment by Jan3k3y** [comment]
When I re-requested the review from kitsunet, Sebobo was unintentionally removed from the reviewers....
But I can not add him again

**Comment by Jan3k3y** [comment]
Does anyone have time for another review?

**Comment by ahaeslich** [comment]
@kitsunet Do we want to include this in the 8.3 release?

**Review by jonnitto — APPROVED** [review]

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (kitsunet) [inline Neos.Media/Classes/Command/MediaCommandController.php]:
```suggestion
                exit(1);
```

I think this should stay, it is an unsuccessful run if nothing was deleted after all?

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (kitsunet) [inline Neos.Media/Classes/Command/MediaCommandController.php]:
```suggestion
                exit(1);
```

Same as above.

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (kitsunet) [inline Neos.Media/Classes/Command/MediaCommandController.php]:
```suggestion
                exit(1);
```

Same as above.

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (kitsunet) [inline Neos.Media/Classes/Command/MediaCommandController.php]:
```suggestion
            exit(1);
```

Same as above. Maybe debatable here and I am fine either way as the user cannot know if presets exist or not, so this might be a valid early exit condition.

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (kitsunet) [inline Neos.Media/Classes/Command/MediaCommandController.php]:
```suggestion
            exit(1);
```

Same as above.

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (kitsunet) [inline Neos.Media/Classes/Command/MediaCommandController.php]:
```suggestion
            exit(1);
```

Same as above.

**Review by kitsunet — APPROVED** [review]
Minor change requests/questions. But wouldn't utlimately block merging.

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (bwaidelich) [reply Neos.Media/Classes/Command/MediaCommandController.php]: why not `$this->quit(1);`?

Comment (kitsunet) [reply Neos.Media/Classes/Command/MediaCommandController.php]: or that, sure!

**Comment by Jan3k3y** [comment]
Should I change it directly in this MR or create a new one and change it in `removeUnusedCommand` too?

**Comment by crydotsnake** [comment]
> Should I change it directly in this MR or create a new one and change it in `removeUnusedCommand` too?

If you mean the suggestions from @kitsunet you can change it in this PR.

**Comment by crydotsnake** [comment]
We should also not forget, to add the new command to the Command Reference in the documentation:

https://github.com/neos/neos-development-collection/blob/8.3/Neos.Neos/Documentation/References/CommandReference.rst#package-neosmedia

**Review by mficzel — DISMISSED** [review]
Just to document the discusssion on the neos sprint:

The complex command choud be split into two commands which would make it more versatile and easy to review. 

One command would allow to show stats about existing variantPresets with an option to highlight/filter those that exist but are not in the configuration. 

The second command would clear the variants of the given preset.

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (mficzel) [inline Neos.Media/Classes/Command/MediaCommandController.php]: why `die`?

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (mficzel) [inline Neos.Media/Classes/Command/MediaCommandController.php]: Would it make sense to have a distinct `removeCustomCropVariantsCommand` instead of setting id and name to `NULL`.

**Review by mficzel — COMMENTED** [review]
I generally like this now :-) cannot really test as i have no case for variants here. 

btw. What will happen if i remove a custom crop that is used.

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (ahaeslich) [reply Neos.Media/Classes/Command/MediaCommandController.php]: Yes, I would also split this in a distinct command as `removeoutdatedvariants` implies that a subset of variants is cleaned. Using `--identifier NULL --variant-name NULL` would AFAIK delete *all* custom crops.

That sounds kind of harmful for me.

Additionally we should split this use case in another issue and not add it on top of this changes.

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (Jan3k3y) [reply Neos.Media/Classes/Command/MediaCommandController.php]: That is a pretty good question 😅

**Comment by Jan3k3y** [comment]
@mficzel 
> btw. What will happen if i remove a custom crop that is used.

The short form: 
It will be removed

The long form: 
If you remove a not configured variant this case is not possible (you can only use configured variants).
If you remove a configured variant there will be an error at the location it would be rendered (don't know the error message but there will be one). To fix this you have to regenerate them with `./flow media:renderVariants`.

Do you think I should ask/inform the user in case a configured variant will be deleted or give the advise to use the `./flow media:renderVariants` command after execution?

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (Jan3k3y) [reply Neos.Media/Classes/Command/MediaCommandController.php]: Removed, but saved for a separat command 😉

Comment (ahaeslich) [reply Neos.Media/Classes/Command/MediaCommandController.php]: Addition: Is the function name still correct? You can now remove variants that are still configured.

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (ahaeslich) [inline Neos.Media/Classes/Command/MediaCommandController.php]:
```suggestion
            $this->outputLine('There is <b>no configuration</b> for the following <b>variant presets:</b>');
```

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (ahaeslich) [inline Neos.Media/Classes/Command/MediaCommandController.php]: I would add `identifier` and `variantName` to the table headers as they are needed in the remove command and it is not obvious which is which. (Also relevant on line 459)

File: Neos.Media/Classes/Command/MediaCommandController.php
Comment (ahaeslich) [inline Neos.Media/Classes/Command/MediaCommandController.php]: This description is not correct anymore as you can also delete configured variants.

**Review by ahaeslich — CHANGES_REQUESTED** [review]

**Comment by mhsdesign** [comment]
Hi ;) what is the status on this? It seems some things changed for the better after we reviewed it at the sprint (eg. we have two commands instead of one big command - i do really like this way ^^)

are there any leftovers to be discussed? or any critical things like: upsi this command deletes wayyyyyy to much?

**Review by mhsdesign — APPROVED** [review]
reviewed, tested and refactored with @ahaeslich bitteschön

**Review by ahaeslich — APPROVED** [review]

**Comment by mhsdesign** [comment]
to get tests working, it could be rebased

File: Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php
Comment (mhsdesign) [inline Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php]: hold my beer, why are the parameters ?nullable

File: Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php
Comment (mhsdesign) [inline Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php]: i think $configuredPresets should not be a dependency as parameter but rather 

```
$currentPresets = $this->assetVariantGenerator->getVariantPresets();
```

should be used

File: Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php
Comment (mhsdesign) [reply Neos.Media/Classes/Domain/Repository/ImageVariantRepository.php]: its not used with nullable params, so we can remove the nullablity and also below: `if (is_null($identifier) && is_null($variantName)) {`

**Review by mhsdesign — CHANGES_REQUESTED** [review]
send @ahaeslich a patch ;)

**Comment by ahaeslich** [comment]
> send @ahaeslich a patch ;)

Thx for the patch ❤️ Checked and pushed it

**Review by mhsdesign — APPROVED** [review]

**Pull request merged**
